src/blockchain/log-processors/order-filled.ts

```diff
@@ -1,17 +1,167 @@
+  db.batchInsert("positions", positionInMarket.map((numShares: string, outcome: number): PositionsRow => ({
+    account,
+    marketID,
+    outcome,
+    numShares,
+    realizedProfitLoss: realizedProfitLoss[outcome],
+    unrealizedProfitLoss: unrealizedProfitLoss[outcome],
+    numSharesAdjustedForUserIntention: positionInMarketAdjustedForUserIntention[outcome],
+  })), 1000).transacting(trx).asCallback(callback);
+}
+
+export function updatePositionInMarket(db: Knex, trx: Knex.Transaction, account: Address, marketID: Address, positionInMarket: Array<string>, realizedProfitLoss: Array<string>, unrealizedProfitLoss: Array<string>, positionInMarketAdjustedForUserIntention: Array<string>, callback: ErrorCallback): void {
+  forEachOf(positionInMarket, (numShares: string, outcome: number, nextOutcome: AsyncCallback): void => {
+    db("positions").transacting(trx).where({ account, marketID, outcome }).update({
+      numShares,
+      realizedProfitLoss: realizedProfitLoss[outcome],
+      unrealizedProfitLoss: unrealizedProfitLoss[outcome],
+      numSharesAdjustedForUserIntention: positionInMarketAdjustedForUserIntention[outcome],
+    }).asCallback(nextOutcome);
+  }, callback);
+}
+
+export function calculateProfitLossInOutcome(augur: Augur, trx: Knex.Transaction, account: Address, marketID: Address, outcome: number, callback: (err: Error|null, profitLossInOutcome?: any) => void): void {
+  getUserTradingHistory(trx, account, marketID, outcome, null, null, null, null, null, (err: Error|null, userTradingHistory?: Array<UITrade>): void => {
+    if (err) return callback(err);
+    trx.select("price").from("outcomes").where({ marketID, outcome }).asCallback((err: Error|null, outcomesRows?: Array<OutcomesRow>): void => {
+      if (err) return callback(err);
+      const { price } = outcomesRows![0];
+      callback(null, augur.trading.calculateProfitLoss({ trades: userTradingHistory, lastPrice: price }));
+    });
+  });
+}
+
+export function upsertPositionInMarket(db: Knex, augur: Augur, trx: Knex.Transaction, account: Address, marketID: Address, positionInMarket: Array<string>, callback: ErrorCallback): void {
+  trx.select("outcome").from("positions").where({ account, marketID }).asCallback((err: Error|null, positionsRows?: Array<PositionsRow>): void => {
+    if (err) return callback(err);
+    const numOutcomes = positionInMarket.length;
+    const realizedProfitLoss = new Array(numOutcomes);
+    const unrealizedProfitLoss = new Array(numOutcomes);
+    const positionInMarketAdjustedForUserIntention = new Array(numOutcomes);
+    forEachOf(positionInMarket, (numShares: string, outcome: number, nextOutcome: AsyncCallback): void => {
+      calculateProfitLossInOutcome(augur, trx, account, marketID, outcome, (err: Error|null, profitLossInOutcome?: any): void => {
```

Comment:
`profitLossInOutcome` could benefit from a type

Reply:
True -- this should be an augur.js type actually, I can add that.